balancer/leastrequest/balancer_test.go

```diff
@@ -287,55 +280,38 @@ func (s) TestLeastRequestE2E(t *testing.T) {
-	var peer1 peer.Peer
-	stream1, err := testServiceClient.FullDuplexCall(ctx, grpc.Peer(&peer1))
-	if err != nil {
-		t.Fatalf("testServiceClient.FullDuplexCall failed: %v", err)
-	}
+		0, 1, // Compares first address (one RPC) to second (no RPCs), so choose second.
+		1, 2, // Compares second address (one RPC) to third (no RPCs), so choose third.
+		0, 3, // Causes another stream on first address.
+		1, 0, // Compares second address (one RPC) to first (two RPCs), so choose second.
+		2, 0, // Compares third address (one RPC) to first (two RPCs), so choose third.
+		0, 0, // Causes another stream on first address.
+		2, 2, // Causes a stream on third address.
+		2, 1, // Compares third address (three RPCs) to second (two RPCs), so choose third.
+	}
+	wantIndex := []uint32{0, 1, 2, 0, 1, 2, 0, 2, 1}
 
-	// Start a second streaming call. From the indexes injected into random
-	// number generator, this should compare Address 1, which already has a RPC
-	// to Address 2, so thus should start the stream on Address 2.
-	var peer2 peer.Peer
-	stream2, err := testServiceClient.FullDuplexCall(ctx, grpc.Peer(&peer2))
-	if err != nil {
-		t.Fatalf("testServiceClient.FullDuplexCall failed: %v", err)
-	}
-
-	// Start a third streaming call. From the indexes injected into random
-	// number generator, this should compare Address 2, which already has a RPC
-	// to Address 3, so thus should start the stream on Address 3.
-	var peer3 peer.Peer
-	stream3, err := testServiceClient.FullDuplexCall(ctx, grpc.Peer(&peer3))
-	if err != nil {
-		t.Fatalf("testServiceClient.FullDuplexCall failed: %v", err)
-	}
-
-	// Finish the streams to collect peer information.
-	stream1.CloseSend()
-	if _, err = stream1.Recv(); err != io.EOF {
-		t.Fatalf("unexpected error: %v, expected an EOF error", err)
-	}
-	stream2.CloseSend()
-	if _, err = stream2.Recv(); err != io.EOF {
-		t.Fatalf("unexpected error: %v, expected an EOF error", err)
-	}
-	stream3.CloseSend()
-	if _, err = stream3.Recv(); err != io.EOF {
-		t.Fatalf("unexpected error: %v, expected an EOF error", err)
-	}
-	if peer1.Addr.String() != peerAtIndex[0] {
-		t.Fatalf("got: %v, want: %v", peer1.Addr.String(), peerAtIndex[0])
-	}
-	if peer2.Addr.String() != peerAtIndex[1] {
-		t.Fatalf("got: %v, want: %v", peer2.Addr.String(), peerAtIndex[1])
-	}
-	if peer3.Addr.String() != peerAtIndex[2] {
-		t.Fatalf("got: %v, want: %v", peer3.Addr.String(), peerAtIndex[2])
+	// Start streaming RPC's, but do not finish them. Each created stream should
+	// be started based on the least request algorithm and injected randomness
+	// (see indexes slice above for exact expectations).
+	for _, wantIndex := range wantIndex {
+		stream, err := testServiceClient.FullDuplexCall(ctx)
+		if err != nil {
+			t.Fatalf("testServiceClient.FullDuplexCall failed: %v", err)
+		}
+		defer func() {
+			stream.CloseSend()
+			if _, err = stream.Recv(); err != io.EOF {
+				t.Fatalf("unexpected error: %v, expected an EOF error", err)
+			}
+		}()
```

Comment:
I think we can just delete all of this and rely on ctx's cancelation to clean up everything.

Same with the other test.  Just make the server's handler do a `<-stream.Context().Done()` instead of doing a `stream.Recv()` in a loop.

Reply:
Done.